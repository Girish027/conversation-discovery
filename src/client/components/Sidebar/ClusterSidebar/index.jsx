import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Popover from 'react-tiny-popover';
import _ from 'lodash';
import { List, Map, fromJS } from 'immutable';
import memoizeOne from 'memoize-one';
import FilterTopic from './FilterTopics';
import SearchBox from '../../SearchBox';
import '../../../styles/CollapsibleOuter.scss';
import {
  Filter,
  ContextualActionsBar,
  ContextualActionItem
} from '@tfs/ui-components';
import ClustersList from './ClusterList';
import {
  clusterSelectors,
  clusterActions,
  initialFilterState,
} from 'state/clusterState';
import { runSelectors } from 'state/runsState';

import Constants from 'components/../constants';
import { saveStorage } from '../../../utils/storageManager';
import { sidebarButtonStyle, popoverContainerStyle } from '../../../styles';
import constants from '../../../constants';
import { prepareClusterName } from '../../../utils/stringOperations';

const filterByFinalizeAndOpen = (cluster, filters) => {
  const initialFilters = initialFilterState;
  if (_.isEqual(filters, initialFilters)) {
    return true;
  }
  const isFinalized = cluster.finalized === filters.finalized;
  const isOpen = !cluster.finalized === filters.open;

  return isFinalized || isOpen;
};

const getMatchedClusters = (clusters, searchString) => {
  let clustersList = clusters.toJS();
  const search = searchString.trim().toLowerCase();

  if (search.length > 0) {
    clustersList = clustersList.filter((cluster) => cluster.rollupCluster.toLowerCase().includes(search)
        || prepareClusterName(cluster.clusterName).toLowerCase().includes(search)
    );
  }
  return clustersList;
};

function getFilteredClusters(clusters, count) {
  return [...clusters].sort((a, b) => (parseInt(a.count, 10) < parseInt(b.count, 10) ? 1 : -1)).splice(0, count);
}

function filterBySimilarityAndVolume(filteredClusters, filters, volume, clusters) {
  if (!_.volume) {
    if ({}.hasOwnProperty.call(volume, constants.filterClusterByVolume.volumeByCount) && volume.volumeByCount !== null) {
      filteredClusters = getFilteredClusters(clusters, volume.volumeByCount);
    } else if ({}.hasOwnProperty.call(volume, constants.filterClusterByVolume.volumeByPercent) && volume.volumeByPercent !== null) {
      const topClusters = Math.round([...clusters].length * volume.volumeByPercent / 100);
      filteredClusters = getFilteredClusters(clusters, topClusters);
    }
  }

  filteredClusters = filteredClusters.filter((cluster) => cluster.similarityCutOff >= filters.similarityCutOff);

  return filteredClusters;
}

const getGroupedClusters = memoizeOne((clustersImmutable = List([]), filtersImmutable = List([]), dispatch) => {
  const clusters = clustersImmutable;
  const filters = filtersImmutable.toJS();
  const newClusterList = [];

  const groupedCluster = [];
  let filteredClusters = [...clusters];

  const { volume } = filters;

  filteredClusters = filterBySimilarityAndVolume(filteredClusters, filters, volume, clusters);

  filteredClusters.forEach((val) => {
    const show = filterByFinalizeAndOpen(val, filters);
    if (show) {
      const hasCluster = groupedCluster.find((cluster) => cluster.rollupCluster === val.rollupCluster);
      if (hasCluster) {
        hasCluster.clusters.push(val);
        hasCluster.count += val.count;
      } else {
        groupedCluster.push({
          rollupCluster: val.rollupCluster,
          count: val.count,
          clusters: [val],
        });
      }
      newClusterList.push(val);
      dispatch(clusterActions.setFilteredClusters(newClusterList));
    }
  });
  return groupedCluster;
});

export class ClustersSidebar extends Component {
  constructor(props) {
    super(props);
    this.props = props;
    this.onSearch = this.onSearch.bind(this);
    this.onClickExpand = this.onClickExpand.bind(this);
    this.handleSorting = this.handleSorting.bind(this);
    this.onClickFilter = this.onClickFilter.bind(this);
    this.handleClosePopover = this.handleClosePopover.bind(this);
    this.getCollapbibleButton = this.getCollapbibleButton.bind(this);

    this.state = {
      clusters: [],
      searchString: '',
      sortOrder: Constants.sortOrder.none,
      isOpen: false,
      isPopoverOpen: false,
    };
  }

  componentDidMount() {
    const { dispatch } = this.props;
    dispatch(clusterActions.loadFilter());
    dispatch(clusterActions.getAllClusters());
  }

  static getDerivedStateFromProps(props, state) {
    const { clusters: existingClustersList, searchString } = state;
    const { clusters: newClusters, filters: newFilters, dispatch } = props;

    // This logic is for performing search operation on the clusters which returns filtered clusters(matched clusters)
    const newCluster = searchString ? getMatchedClusters(newClusters, searchString) : newClusters.toJS();
    const existingClusters = [];
    const newState = {};

    const initialFilters = initialFilterState;
    const filters = newFilters.toJS();

    // This logic is for comparing the existing clusters with new clusters(when props changes)
    existingClustersList.forEach((cluster) => {
      existingClusters.push(...cluster.clusters);
    });
    newCluster.sort((a, b) => a.clusterId - b.clusterId);
    existingClusters.sort((a, b) => a.clusterId - b.clusterId);

    if (!_.isEqual(existingClusters, newCluster) || !_.isEqual(filters, initialFilters)) {
      newState.clusters = getGroupedClusters(newCluster, newFilters, dispatch);
    }
    return newState;
  }

  onSearch(searchString) {
    this.setState({
      searchString
    });
  }

  onClickExpand() {
    this.setState({
      isOpen: !this.state.isOpen
    });
  }

  onClickFilter() {
    this.setState({
      isPopoverOpen: !this.state.isPopoverOpen
    });
  }

  getCollapbibleButton(isOpen) {
    let buttonClass = 'CollapsibleOuter';
    if (isOpen) {
      buttonClass = 'CollapsibleOuter isOpen';
    }

    return (
      <div>
        <span className={buttonClass}>
          {!isOpen
            ? Constants.buttons.expandAll
            : Constants.buttons.collapseAll
          }
        </span>
      </div>
    );
  }

  handleSorting() {
    const { sortOrder } = this.state;
    if (sortOrder !== Constants.sortOrder.asc) {
      this.setState({
        sortOrder: Constants.sortOrder.asc,
      });
      saveStorage(Constants.clusterSort, { sortOrder: Constants.sortOrder.asc });
    } else {
      this.setState({
        sortOrder: Constants.sortOrder.desc,
      });
      saveStorage(Constants.clusterSort, { sortOrder: Constants.sortOrder.desc });
    }
  }

  handleClosePopover() {
    this.setState({
      isPopoverOpen: false
    });
  }

  render() {
    const {
      isOpen, sortOrder, isPopoverOpen, clusters
    } = this.state;

    return (
      <div>
        <div style={{ display: 'flex' }}>
          <SearchBox
            onSearch={this.onSearch}
            placeholder='Find Intents'
            styleOverride={{ width: '90%' }}
          />

          <Popover
            isOpen={isPopoverOpen}
            position={'bottom'}
            align={'start'}
            onClickOutside={this.handleClosePopover}
            containerStyle={popoverContainerStyle}
            content={(
              <FilterTopic />
            )}
          >
            <ContextualActionsBar styleOverride={{ ...sidebarButtonStyle.containerStyle, width: '48%' }}>
              <ContextualActionItem
                right
                icon={Filter}
                onClickAction={this.onClickFilter}
                styleOverride={{ fontSize: '12px', marginRight: '-7px', marginTop: '23px' }}
              >
                FILTER
              </ContextualActionItem>
            </ContextualActionsBar>
          </Popover>
        </div>
        <ClustersList
          isOpen={isOpen}
          clusters={clusters}
          handleSorting={this.handleSorting}
          sortOrder={sortOrder}
        />
      </div>
    );
  }
}

ClustersSidebar.propTypes = {
  dispatch: PropTypes.func,
  // eslint-disable-next-line react/no-unused-prop-types
  filters: PropTypes.object,
};

ClustersSidebar.defaultProps = {
  dispatch: () => {},
  filters: fromJS(initialFilterState),
};

const mapStateToProps = (state) => {
  const allFilters = clusterSelectors.getFilters(state) || new Map({});
  const selectedRunId = runSelectors.getSelectedRunId(state) || '';
  const filters = allFilters.get(selectedRunId);
  return {
    clusters: clusterSelectors.getClusterList(state) || new List([]),
    filters,
  };
};

const mapDispatchToProps = (dispatch) => ({
  dispatch,
});


export default connect(mapStateToProps, mapDispatchToProps)(ClustersSidebar);
