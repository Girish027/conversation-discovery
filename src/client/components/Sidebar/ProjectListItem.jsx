import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  Xmark,
  Pencil,
  StatusBadge,
} from '@tfs/ui-components';
import IconButton from 'components/IconButton';
import { getDateFromTimestamp } from 'utils/DateUtils';
import { colors } from 'styles';
import { shouldHandle } from 'utils/KeyboardUtils';
import Constants from '../../constants';
import Language from '../../Language';

import { projectActions } from '../../state/projectsState';
import {
  setModalIsOpen,
} from 'state/appState';

class ProjectListItem extends Component {
  constructor() {
    super();

    this.renderDeleteIcon = this.renderDeleteIcon.bind(this);
    this.renderEditIcon = this.renderEditIcon.bind(this);

    this.toggleFocus = this.toggleFocus.bind(this);
    this.onMouseUp = this.onMouseUp.bind(this);
    this.onClickListItem = this.onClickListItem.bind(this);
    this.onClickEdit = this.onClickEdit.bind(this);
    this.onClickDelete = this.onClickDelete.bind(this);
    this.onKeyDown = this.onKeyDown.bind(this);
    this.renderSystemProjItemList = this.renderSystemProjItemList.bind(this);
    this.renderManualProjItemList = this.renderManualProjItemList.bind(this);

    this.state = {
      focus: false,
    };
  }

  onMouseUp() {
    this.setState({
      focus: false,
    });
  }

  onKeyDown(event) {
    if (shouldHandle(event)) {
      this.onClickListItem();
    }
  }

  onClickEdit(event) {
    event.stopPropagation();
    const { dispatch, projectId, ...projectInfo } = this.props;
    dispatch(setModalIsOpen(true, {
      modalName: Constants.modals.editProject,
      header: Constants.modalInfo.editProject.header,
      projectId,
      formData: projectInfo,
    }));
  }

  onClickDelete(event) {
    event.stopPropagation();
    const { dispatch, projectName } = this.props;
    // Ask for confirmation - open Confirmation Dialog
    dispatch(setModalIsOpen(true, {
      modalName: Constants.modals.confirm,
      deleteMode: true,
      header: `Delete "${projectName}"?`,
      message: Language.WARNING_MESSAGE.deleteTranscript,
      onClickOk: this.onClickDeleteButton
    }));
  }

  onClickDeleteButton = () => {
    const { dispatch, projectId } = this.props;
    dispatch(projectActions.deleteProject(projectId));
  }

  onClickListItem() {
    const { projectId, onClickListItem } = this.props;
    onClickListItem(projectId);
  }

  toggleFocus() {
    const { focus } = this.state;
    const { onFocusItem, projectId } = this.props;
    this.setState({
      focus: !focus,
    }, () => {
      onFocusItem(projectId, this.state.focus);
    });
  }

  renderDeleteIcon() {
    return (
      <IconButton
        onClick={this.onClickDelete}
        icon={Xmark}
        data-qa={'delete-project'}
        padding={'3px'}
        title='Delete'
      />
    );
  }

  renderEditIcon() {
    return (
      <IconButton
        onClick={this.onClickEdit}
        icon={Pencil}
        defaultColor='none'
        strokeColor='#004c97'
        data-qa={'edit-project'}
        padding={'3px'}
        title='Edit'
      />
    );
  }

  renderSystemProjItemList() {
    const {
      projectDescription,
      projectStatus,
      modified,
      focused,
      roleConfig,
    } = this.props;

    let badgeCategory = 'primary';
    if (projectStatus === Constants.status.NEW) {
      badgeCategory = 'success';
    }

    return (
      <div>
        <div className='project-name'>
          {Constants.LIVE_DATA_ANALYSIS}
          <StatusBadge
            category={badgeCategory}
            label={projectStatus}
            onHover={() => { }}
            progress='true'
            type='badge'
            styleOverride={{ marginLeft: '20px', paddingTop: '2px' }}
          />
        </div>

        <div className='project-meta'>
          <div style={{ color: '#313f54', marginBottom: '4px' }}>
            {projectDescription}
          </div>
          {`Last Edited: ${getDateFromTimestamp(modified)}`}
        </div>
        {focused && (
          <div className='actions'>
            { roleConfig.deleteTranscripts ? this.renderDeleteIcon() : null }
          </div>
        )}
      </div>
    );
  }

  renderManualProjItemList() {
    const {
      projectName,
      projectDescription,
      modified,
      datasetName,
      focused,
      roleConfig,
    } = this.props;

    return (
      <div>
        <div className='project-name'>
          {projectName}
        </div>

        <div className='project-meta'>
          <div style={{ color: '#313f54' }}>
            {projectDescription}
          </div>
          {`Dataset: ${datasetName}`}
          <br />
          {`Last Edited: ${getDateFromTimestamp(modified)}`}
        </div>
        {focused && (
          <div className='actions'>
            { roleConfig.updateTranscripts ? this.renderEditIcon() : null }
            { roleConfig.deleteTranscripts ? this.renderDeleteIcon() : null }
          </div>
        )}
      </div>
    );
  }

  render() {
    const {
      projectType,
      focused,
      selected,
      sticky,
    } = this.props;
    return (
      <div
        className={`project-list-item ${sticky}`}
        onClick={this.onClickListItem}
        onKeyDown={this.onKeyDown}
        onFocus={this.toggleFocus}
        onBlur={this.toggleFocus}
        onMouseOver={this.toggleFocus}
        onMouseOut={this.toggleFocus}
        onMouseUp={this.onMouseUp}
        role='button'
        tabIndex={0}
        style={{
          backgroundColor: focused ? colors.focusItem : (selected ? colors.selectedItem : ''),
        }}
      >
        { (projectType && projectType === Constants.projects.SYSTEM) ? (
          this.renderSystemProjItemList()
        ) : (
          this.renderManualProjItemList()
        )
        }
      </div>
    );
  }
}

ProjectListItem.propTypes = {
  sticky: PropTypes.string,
  projectName: PropTypes.string.isRequired,
  projectDescription: PropTypes.string,
  projectStatus: PropTypes.string,
  projectType: PropTypes.string,
  projectId: PropTypes.string.isRequired,
  datasetName: PropTypes.string.isRequired,
  modified: PropTypes.number.isRequired,
  focused: PropTypes.bool,
  selected: PropTypes.bool,
  onFocusItem: PropTypes.func,
  onClickListItem: PropTypes.func,
  roleConfig: PropTypes.object,
  dispatch: PropTypes.func,
};

ProjectListItem.defaultProps = {
  projectDescription: '',
  sticky: '',
  focused: false,
  selected: false,
  onFocusItem: () => { },
  onClickListItem: () => { },
  roleConfig: {},
  dispatch: () => {},
};

export default ProjectListItem;
