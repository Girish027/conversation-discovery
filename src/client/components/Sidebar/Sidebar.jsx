import React, { Component } from 'react';
import { Switch, Route } from 'react-router-dom';
import { routeNames } from 'utils/RouteHelper';
import ConnectedProjectsSidebar from './ProjectsSidebar';
import ClustersSidebar from 'components/Sidebar/ClusterSidebar';
import IconButton from 'components/IconButton';
import AngleRight from 'components/Icons/AngleRight';
import AngleLeft from 'components/Icons/AngleLeft';

import 'styles/sidebar.scss';

class Sidebar extends Component {
  constructor(props) {
    super(props);

    this.toggleSidebar = this.toggleSidebar.bind(this);
    this.state = {
      sidebarOpen: true,
    };
  }

  toggleSidebar() {
    const { sidebarOpen } = this.state;
    this.setState({
      sidebarOpen: !sidebarOpen
    });
  }

  render() {
    const { sidebarOpen } = this.state;
    let sidebarClassName = '';
    if (sidebarOpen) {
      sidebarClassName = 'sidebar-open';
    } else {
      sidebarClassName = 'sidebar-closed';
    }
    return (
      <div className={`sidebar-container ${sidebarClassName}`}>
        { sidebarOpen && (
          <Switch>
            <Route exact path={routeNames.DISTRIBUTION_GRAPH} component={ClustersSidebar} />
            <Route exact path={routeNames.TOPIC_REVIEW} component={ClustersSidebar} />
            <Route exact path={routeNames.DISCOVER_INTENTS} component={ConnectedProjectsSidebar} />
            <Route path={routeNames.HOME_PAGE} component={ConnectedProjectsSidebar} />
          </Switch>

        )}
        <div className={`sidebar-toggle ${sidebarClassName}`}>
          { sidebarOpen ? (
            <IconButton
              onClick={this.toggleSidebar}
              icon={AngleLeft}
              title='Hide Sidebar'
            />
          ) : (
            <IconButton
              onClick={this.toggleSidebar}
              icon={AngleRight}
              title='Show Sidebar'
            />
          )}
        </div>
      </div>
    );
  }
}

export default Sidebar;
