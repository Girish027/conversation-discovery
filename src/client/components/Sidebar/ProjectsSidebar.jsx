import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import _ from 'lodash';
import {
  Plus
} from '@tfs/ui-components';
import { List } from 'immutable';
import IconButton from 'components/IconButton';

import {
  setModalIsOpen,
  getCAASelector,
  getClientIdSelector,
  getRoleConfigSelector,
} from 'state/appState';
import {
  projectActions,
  projectSelectors,
} from 'state/projectsState';
import { headerActions } from 'state/headerState';
import * as ValidationUtils from '../../utils/ValidationUtils';

// '../../ValidationUtils';
import Constants from '../../constants';
import ProjectListItem from './ProjectListItem';

export class ProjectsSidebar extends Component {
  constructor(props) {
    super(props);

    this.onClickCreateProject = this.onClickCreateProject.bind(this);
    this.onSelectProject = this.onSelectProject.bind(this);
    this.onFocusProjectItem = this.onFocusProjectItem.bind(this);
    this.clearFocus = this.clearFocus.bind(this);
    this.renderManualProjectList = this.renderManualProjectList.bind(this);
    this.renderSystemProjectList = this.renderSystemProjectList.bind(this);
    this.checkType = this.checkType.bind(this);
    this.getSystemProjectList = this.getSystemProjectList.bind(this);
    this.renderSystemProjectWorkSpace = this.renderSystemProjectWorkSpace.bind(this);
    this.renderManualProjectWorkSpace = this.renderManualProjectWorkSpace.bind(this);

    this.projectTimer = null;

    this.state = {
      caa: {},
      shouldFetchProjects: false,
      activeProjectId: '',
      focusedProjectId: '',
    };
  }

  static getDerivedStateFromProps(props, state) {
    const { caa: existingCAA } = state;
    const { caa: newCAA, activeProjectId } = props;
    const newState = {
      activeProjectId
    };
    if (!_.isEqual(existingCAA, newCAA)) {
      newState.caa = newCAA;
      newState.shouldFetchProjects = true;
    } else {
      newState.shouldFetchProjects = false;
    }
    return newState;
  }

  componentDidUpdate() {
    const { dispatch } = this.props;
    const { shouldFetchProjects, caa } = this.state;
    if (shouldFetchProjects) {
      dispatch(projectActions.getAllProjects(caa));
    }
  }

  onClickCreateProject() {
    const { projects: projectsImmutable, dispatch, clientId } = this.props;
    const projects = projectsImmutable.toJS();
    const numberOfProjects = projects.length;
    const isValidObj = ValidationUtils.createProjectValidation(numberOfProjects, clientId);
    if (isValidObj && !isValidObj.isValid) {
      dispatch(headerActions.showNotificationWithTimer(isValidObj.message, Constants.notification.error));
      return;
    }
    dispatch(setModalIsOpen(true, {
      modalName: Constants.modals.createProject,
    }));
  }

  onFocusProjectItem(projectId, inFocus) {
    if (inFocus) {
      this.setState({
        focusedProjectId: projectId
      });
    }
  }

  onSelectProject(projectId) {
    const { dispatch } = this.props;
    dispatch(projectActions.selectProject(projectId));
  }

  getSystemProjectList() {
    const { projects: projectsImmutable } = this.props;
    const systemProjects = projectsImmutable.toJS();
    return systemProjects.filter((project) => this.checkType(project, Constants.projects.SYSTEM));
  }

  clearFocus() {
    this.setState({
      focusedProjectId: ''
    });
  }

  checkType(project, type) {
    const { projectType } = project;
    if (projectType === type) {
      return true;
    }
    return false;
  }

  renderManualProjectList() {
    const { focusedProjectId, activeProjectId } = this.state;
    const { projects: projectsImmutable, dispatch, roleConfig } = this.props;
    const manualProjects = projectsImmutable.toJS();

    let manualListItems = manualProjects.filter((project) => this.checkType(project, Constants.projects.MANUAL));

    if (manualListItems.length > 0) {
      manualListItems = manualListItems.map((project) => {
        const { projectId, projectName } = project;
        const selected = projectId === activeProjectId;
        return (
          <ProjectListItem
            sticky={selected ? 'sticky' : ''}
            key={`listitem-${projectName}`}
            onClickListItem={this.onSelectProject}
            onFocusItem={this.onFocusProjectItem}
            focused={projectId === focusedProjectId}
            selected={selected}
            dispatch={dispatch}
            roleConfig={roleConfig}
            {...project}
          />
        );
      });
    }

    return (manualListItems);
  }

  renderSystemProjectList(systemListItems) {
    const { focusedProjectId, activeProjectId } = this.state;
    const { dispatch, roleConfig } = this.props;

    systemListItems = systemListItems.map((project) => {
      const { projectId, projectName } = project;
      const selected = projectId === activeProjectId;
      return (
        <ProjectListItem
          sticky={selected ? 'sticky' : ''}
          key={`listitem-${projectName}`}
          onClickListItem={this.onSelectProject}
          onFocusItem={this.onFocusProjectItem}
          focused={projectId === focusedProjectId}
          selected={selected}
          dispatch={dispatch}
          roleConfig={roleConfig}
          {...project}
        />
      );
    });

    return (systemListItems);
  }

  renderSystemProjectWorkSpace(systemListItems) {
    return (
      <div className='sidebar-project'>
        <div className='title sticky'>
          <span className='float-left-project'>
            Automated Discovery
          </span>
          <span className='float-clear' />
        </div>
        <div
          className='auto-project-list system-project-list-layout'
          onMouseLeave={this.clearFocus}
        >
          {this.renderSystemProjectList(systemListItems)}
        </div>
      </div>
    );
  }

  renderManualProjectWorkSpace(autoProjNumber) {
    const { roleConfig } = this.props;
    let styleListHeight = '';
    let titleStyle = 'sub-title sticky';
    if (autoProjNumber === 0) {
      titleStyle = 'title sticky';
      styleListHeight = 'project-list project-list-height';
    } else if (autoProjNumber === 1) {
      styleListHeight = 'project-list project-list-height-1auto';
    } else if (autoProjNumber === 2) {
      styleListHeight = 'project-list project-list-height-2auto';
    }

    return (
      <div className='sidebar-project'>
        <div className={titleStyle}>
          <span className='float-left-project'>
            Manual Discovery
          </span>
          <span className='float-right'>
            { roleConfig.uploadNewTranscript ? (
              <IconButton
                onClick={this.onClickCreateProject}
                icon={Plus}
                strokeColor='#ffffff'
                title='Create Project'
              />
            ) : null }
          </span>
          <span className='float-clear' />
        </div>
        <div
          className={styleListHeight}
          onMouseLeave={this.clearFocus}
        >
          {this.renderManualProjectList()}
        </div>
      </div>
    );
  }

  render() {
    const systemListItems = this.getSystemProjectList();
    return (
      <div>
        {(systemListItems.length > 0) && this.renderSystemProjectWorkSpace(systemListItems)}
        {this.renderManualProjectWorkSpace(systemListItems.length)}
      </div>
    );
  }
}

ProjectsSidebar.propTypes = {
  activeProjectId: PropTypes.string,
  caa: PropTypes.object.isRequired,
  dispatch: PropTypes.func,
  projects: PropTypes.object,
  clientId: PropTypes.string,
  roleConfig: PropTypes.object,
};

ProjectsSidebar.defaultProps = {
  dispatch: () => { },
  activeProjectId: '',
  projects: new List([]),
  roleConfig: {},
};

const mapStateToProps = (state) => ({
  activeProjectId: projectSelectors.getActiveProjectId(state) || '',
  caa: getCAASelector(state),
  clientId: getClientIdSelector(state),
  projects: projectSelectors.getProjectsList(state) || new List([]),
  roleConfig: getRoleConfigSelector(state),
});

const mapDispatchToProps = (dispatch) => ({
  dispatch,
});


export default connect(mapStateToProps, mapDispatchToProps)(ProjectsSidebar);
