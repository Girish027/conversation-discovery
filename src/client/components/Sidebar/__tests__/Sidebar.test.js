import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Sidebar from 'components/Sidebar';

describe('<Sidebar />', function () {
  let wrapper;

  const getWrapper = () => shallow(
    <Sidebar />
  );
  describe('Creating an instance', () => {
    test('should exist', () => {
      wrapper = getWrapper();
      expect(wrapper.exists()).toBe(true);
    });
  });

  describe('Snapshots', () => {
    test('renders correctly with routes and is open', () => {
      wrapper = getWrapper();
      expect(toJSON(wrapper)).toMatchSnapshot();
    });

    test('renders correctly when sidebar is closed', () => {
      wrapper = getWrapper();
      wrapper.instance().toggleSidebar();
      expect(toJSON(wrapper)).toMatchSnapshot();
    });

    test('renders correctly when sidebar is closed and then opened', () => {
      wrapper = getWrapper();
      wrapper.instance().toggleSidebar();
      expect(wrapper.state().sidebarOpen).toBe(false);
      wrapper.instance().toggleSidebar();
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });
});
