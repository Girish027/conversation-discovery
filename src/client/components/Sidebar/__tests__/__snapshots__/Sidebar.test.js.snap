// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> Snapshots renders correctly when sidebar is closed 1`] = `
<div
  className="sidebar-container sidebar-closed"
>
  <div
    className="sidebar-toggle sidebar-closed"
  >
    <IconButton
      defaultColor="#004c97"
      disabled={false}
      disabledColor="#9B9B9B"
      focusedColor="#003467"
      height={10}
      icon={[Function]}
      onClick={[Function]}
      padding="1px 1px 3px 1px"
      strokeColor="none"
      title="Show Sidebar"
      width={10}
    />
  </div>
</div>
`;

exports[`<Sidebar /> Snapshots renders correctly when sidebar is closed and then opened 1`] = `
<div
  className="sidebar-container sidebar-open"
>
  <Switch>
    <Route
      component={[Function]}
      exact={true}
      path="/conversationdiscovery/ui/run-overview"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/conversationdiscovery/ui/topic-review"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/conversationdiscovery/ui/discover-intents"
    />
    <Route
      component={[Function]}
      path="/conversationdiscovery"
    />
  </Switch>
  <div
    className="sidebar-toggle sidebar-open"
  >
    <IconButton
      defaultColor="#004c97"
      disabled={false}
      disabledColor="#9B9B9B"
      focusedColor="#003467"
      height={10}
      icon={[Function]}
      onClick={[Function]}
      padding="1px 1px 3px 1px"
      strokeColor="none"
      title="Hide Sidebar"
      width={10}
    />
  </div>
</div>
`;

exports[`<Sidebar /> Snapshots renders correctly with routes and is open 1`] = `
<div
  className="sidebar-container sidebar-open"
>
  <Switch>
    <Route
      component={[Function]}
      exact={true}
      path="/conversationdiscovery/ui/run-overview"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/conversationdiscovery/ui/topic-review"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/conversationdiscovery/ui/discover-intents"
    />
    <Route
      component={[Function]}
      path="/conversationdiscovery"
    />
  </Switch>
  <div
    className="sidebar-toggle sidebar-open"
  >
    <IconButton
      defaultColor="#004c97"
      disabled={false}
      disabledColor="#9B9B9B"
      focusedColor="#003467"
      height={10}
      icon={[Function]}
      onClick={[Function]}
      padding="1px 1px 3px 1px"
      strokeColor="none"
      title="Hide Sidebar"
      width={10}
    />
  </div>
</div>
`;
